app/src/main/java/edu/uw/myuw_android/LoginActivity.kt

```diff
@@ -0,0 +1,138 @@
+
+class LoginActivity: AppCompatActivity() {
+
+    val RC_AUTH = 132
+    lateinit var authorizationService: AuthorizationService
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_login)
+        supportActionBar?.hide()
+
+        findViewById<Button>(R.id.button).setOnClickListener {
+            tryLoginWithAppAuth()
+        }
+
+        authorizationService = AuthorizationService(this)
+
+        if (UserInfoStore.readAuthState(this).isAuthorized) {
+            findViewById<TextView>(R.id.textView).text = "You are Authorized"
+            findViewById<Button>(R.id.button).isClickable = false
+            startMainActivity()
+        } else {
+            findViewById<TextView>(R.id.textView).text = "You are not Authorized"
+        }
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        authorizationService.dispose()
+    }
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        if (requestCode == RC_AUTH) {
+            val resp = AuthorizationResponse.fromIntent(data!!)
+            val ex = AuthorizationException.fromIntent(data)
+
+            if (resp != null) {
+                val authState = AuthState(resp, ex)
+                authorizationService.performTokenRequest(
+                    resp.createTokenExchangeRequest()
+                ) { response, ex_new ->
+                    if (response != null) {
+                        authState.update(response, ex)
+                        UserInfoStore.writeAuthState(this, authState)
+                        startMainActivity()
+                    } else {
+                        Log.e("performTokenRequest", ex_new!!.localizedMessage!!)
+                    }
+                }
+            } else {
+                Log.e("AuthorizationResponse", ex!!.localizedMessage!!)
+            }
+        } else {
+            super.onActivityResult(requestCode, resultCode, data)
+        }
+    }
+
+    private fun tryLoginWithAppAuth() {
+        AuthorizationServiceConfiguration.fetchFromUrl(
+            Uri.parse(resources.getString(R.string.openid_discovery_uri))
+        ) { serviceConfiguration, ex ->
+            if (serviceConfiguration != null) {
+                val authRequestBuilder = AuthorizationRequest.Builder(
+                    serviceConfiguration,
+                    resources.getString(R.string.openid_client_id),
+                    ResponseTypeValues.CODE,
+                    Uri.parse(resources.getString(R.string.openid_redirect_uri))
+                )
+
+                val authRequest = authRequestBuilder
+                    .setScope(resources.getString(R.string.openid_authorization_scope))
+                    .build()
+
+                startActivityForResult(
+                    authorizationService.getAuthorizationRequestIntent(authRequest),
+                    RC_AUTH
+                )
```

Comment:
I think it will be extremely hard to write test around this function because:
1. The line 85, 89, 96 and 101 are appauth library calls that have been tested by appauth.
2. The intent returned by getAuthorizationRequestIntent can only be validated by the oidc provider. The structure is guaranteed and tested by appauth.
3. The only input to this function is configuration. We can do very little to validate the configuration as any url string in it would look like a valid config but we can never be sure till the oidc client replies.